### PR TITLE
Add OpsGenie Receiver type to AlertmanagerConfig CRD

### PR DESCRIPTION
**Pull request by omegas27**
Towards https://github.com/prometheus-operator/prometheus-operator/issues/3536

### CODE REVIEW

**Comment by brancz** [comment]
cc @grdryn

File: pkg/alertmanager/amcfg.go [hunk 1]
```diff
@@ -294,9 +317,18 @@ func (cg *configGenerator) convertReceiver(ctx context.Context, in *monitoringv1
 		}
 	}
 
+	var opsgenieConfigs []*opsgenieConfig
+	if l := len(in.OpsGenieConfigs); l > 0 {
+		opsgenieConfigs = make([]*opsgenieConfig, l)
+		for i := range in.OpsGenieConfigs {
+			opsgenieConfigs[i] = cg.convertOpsgenieConfig(ctx, in.OpsGenieConfigs[i], crKey)
```
Comment (paulfantom) [inline pkg/alertmanager/amcfg.go:324]: Let's wait for https://github.com/prometheus-operator/prometheus-operator/pull/3604 and after it gets merged reafctor this to handle incorrect opsgenie configs.

Comment (omegas27) [reply pkg/alertmanager/amcfg.go]: I just rebased the master branch as this got merged

File: pkg/alertmanager/operator.go [hunk 2]
```diff
@@ -935,6 +936,24 @@ func (c *Operator) selectAlertManagerConfigs(ctx context.Context, am *monitoring
 	return res, nil
 }
 
+// configureHTTPConfigInStore configure the asset store for HTTPConfigs.
+func configureHTTPConfigInStore(ctx context.Context, store *assetStore, httpConfig *monitoringv1alpha1.HTTPConfig, amcNamespace string, receiverKey string) {
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:940]: Let's make it a method of `assetStore`. And it should return an `error`.

```suggestion
func (a *assetStore) configureHTTPConfigInStore(ctx context.Context, httpConfig *monitoringv1alpha1.HTTPConfig, namespace string, key string) error {
```

File: pkg/alertmanager/operator.go [hunk 3]
```diff
@@ -881,22 +882,22 @@ func (c *Operator) selectAlertManagerConfigs(ctx context.Context, am *monitoring
...
 				}
 
-				if pdConfig.HTTPConfig == nil {
-					continue
+				if pdConfig.HTTPConfig != nil {
+					configureHTTPConfigInStore(ctx, store, pdConfig.HTTPConfig, amcNamespace, pdcKey)
 				}
+			}
+
+			for j, ogConfig := range receiver.OpsGenieConfigs {
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:890]: Alertmanager requires that at least one of `id`, `name` or `username` is not empty. The value of `type` should also match `^(team|user|escalation|schedule)$`. We need to enforce it here.

Comment (omegas27) [reply pkg/alertmanager/operator.go:890]: Would the prefered behavior be to skip the responder that is not matching here or the whole opsgenie config?

File: pkg/alertmanager/operator.go (hunk 2 above)
Comment (omegas27) [reply pkg/alertmanager/operator.go:940]: Thanks that looks way better

File: pkg/alertmanager/operator.go (hunk 3 above)
Comment (simonpasquier) [reply pkg/alertmanager/operator.go:890]: I would skip the whole config.

Comment (omegas27) [reply pkg/alertmanager/operator.go:892]: Thanks. The latest commit should address your comment :D

File: pkg/alertmanager/operator.go [hunk 4]
```diff
@@ -881,22 +882,32 @@ func (c *Operator) selectAlertManagerConfigs(ctx context.Context, am *monitoring
 					}
 				}
 
-				if pdConfig.HTTPConfig == nil {
-					continue
+				if pdConfig.HTTPConfig != nil {
+					if err = store.configureHTTPConfigInStore(ctx, pdConfig.HTTPConfig, amcNamespace, pdcKey); err != nil {
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:886]: I would update `configureHTTPConfigInStore()` to accept `nil` values for `*HTTPConfig` (and return early) as it would remove the need to test for `nil` everywhere.

File: pkg/alertmanager/operator.go [hunk 5]
```diff
@@ -881,23 +882,30 @@ func (c *Operator) selectAlertManagerConfigs(ctx context.Context, am *monitoring
...
 						break
 					}
 				}
 
-				if err = store.addBasicAuth(ctx, amc.GetNamespace(), pdConfig.HTTPConfig.BasicAuth, pdcKey); err != nil {
-					break
+				if len(ogConfig.Responders) > 0 {
+					if err = ogConfig.Validate(); err != nil {
+						break
+					}
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:902]:
```suggestion
				if err = ogConfig.Validate(); err != nil {
					break
				}
```

**Review by simonpasquier — COMMENTED** [review]
You'd need to rebase because of #3606.

**Comment by simonpasquier** [comment]
@omegas27 you'd need to rebase once again. Sorry about this but the PR looks great. I'm holding a bit the next release so it can include this change.

**Comment by omegas27** [comment]
@simonpasquier  No worries, I just rebased

**Review by simonpasquier — APPROVED** [review]
Thanks!
